cmd/backup/configProvider.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021-2022 - Offen Authors <hioffen@posteo.de>
+// SPDX-License-Identifier: MPL-2.0
+
+package main
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/joho/godotenv"
+	"github.com/offen/envconfig"
+)
+
+type envProxy func(string) (string, bool)
+
+func loadConfig(lookup envProxy) (*Config, error) {
+	envconfig.Lookup = func(key string) (string, bool) {
+		value, okValue := lookup(key)
+		location, okFile := lookup(key + "_FILE")
+
+		switch {
+		case okValue && !okFile: // only value
+			return value, true
+		case !okValue && okFile: // only file
+			contents, err := os.ReadFile(location)
+			if err != nil {
+				slog.Error(fmt.Sprintf("failed to read %s", location), "error", err)
+				return "", false
+			}
+			return string(contents), true
+		case okValue && okFile: // both
+			slog.Error(fmt.Sprintf("both %s and %s are set!", key, key+"_FILE"))
+			return "", false
+		default: // neither, ignore
+			return "", false
+		}
+	}
+
+	var c = &Config{}
+	if err := envconfig.Process("", c); err != nil {
+		slog.Error("failed to process configuration values", "error", err)
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func loadEnvVars() (*Config, error) {
+	slog.Info("loading config from environment variables")
+	return loadConfig(os.LookupEnv)
+}
+
+func loadEnvFiles(folder string) ([]*Config, error) {
+	slog.Info(fmt.Sprintf("loading config from environment files from directory %s", folder))
+
+	items, err := os.ReadDir(folder)
```

Comment:
This is checked outside the call. The `loadEnvFiles` function tries to load from a location. If it does not exist, it will report that. It is up to the caller to decide if that error is recoverable.

Reply:
I understand that, but from reading the docs, I am not sure whether `os.ReadDir` will ever return such an error

> ReadDir reads the named directory, returning all its directory entries sorted by filename. If an error occurs reading the directory, ReadDir returns the entries it was able to read before the error, along with the error.

which is why I thought we'd need to cater for the case of an inexistent directory ourselve?.

Reply:
My understanding is that the folder not existing would certainly trigger an error during reading and hence it would be returned by `os.ReadDir`. It is also the behaviour that I observed.

Reply:
Ah, I did not know that `os.IsNotExist` also picks up syscall errors:

```
// IsNotExist returns a boolean indicating whether the error is known to
// report that a file or directory does not exist. It is satisfied by
// ErrNotExist as well as some syscall errors.
```

I was thinking it's just really checking for `os.ErrNotExist`. But in this case this should be working as is. Sorry for the confusion.